ledger/complete/ledger.go

```diff
@@ -86,19 +89,64 @@ func NewLedger(
 	return storage, nil
 }
 
+func NewSyncLedger(
+	wal realWAL.LedgerWAL,
+	capacity int,
+	metrics module.LedgerMetrics,
+	log zerolog.Logger,
+	pathFinderVer uint8) (*Ledger, error) {
+
+	l, err := NewLedger(wal, capacity, metrics, log, pathFinderVer)
+	if err != nil {
+		return nil, err
+	}
+
+	l.trieUpdateCh = make(chan *WALTrieUpdate, defaultTrieUpdateChanSize)
+	l.trieUpdateDoneCh = make(chan struct{})
+	return l, nil
+}
+
+// TrieUpdateChan returns a channel which is used to receive trie updates that needs to be logged in WALs.
+// This channel is closed when ledger component shutdowns down.
+func (l *Ledger) TrieUpdateChan() <-chan *WALTrieUpdate {
+	return l.trieUpdateCh
+}
+
+// TrieUpdateDoneChan returns a channel which is closed when there are no more WAL updates.
+// This is used to signal that it is safe to shutdown WAL component (closing opened WAL segment).
```

Comment:
> Why do we need extra channel? Can we just close the channel to indicate there is no more updates?

We use the extra channel because when trie update channel is closed, Compactor still needs to drain and process remaining trie update in the closed channel.  Ledger is responsible for closing opened WAL segment, so `TrieUpdateDoneChan` is used to indicate all WAL writing activities are done in `Compactor`.

On the other hand, if `Compactor` is responsible for closing opened WAL segment, this extra communication isn't needed.  Thanks for bringing this up :+1: , I'll look into it.